test/e2e/upgrade/alert/alert.go

```diff
@@ -0,0 +1,224 @@
+package alert
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+	"github.com/prometheus/common/model"
+
+	v1 "k8s.io/api/core/v1"
+	kapierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	watchtools "k8s.io/client-go/tools/watch"
+	"k8s.io/kubernetes/pkg/client/conditions"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/upgrades"
+)
+
+const (
+	alertCheckSleep            = 5 * time.Minute
+	maxPrometheusQueryAttempts = 5
+	prometheusQueryRetrySleep  = 10 * time.Second
+)
+
+// UpgradeTest tests that a service is available before, during, and
+// after a cluster upgrade.
+type UpgradeTest struct {
+	url         string
+	bearerToken string
+	oc          *exutil.CLI
+}
+
+func (UpgradeTest) Name() string { return "check-for-critical-alerts" }
+func (UpgradeTest) DisplayName() string {
+	return "Check if critical alerts exist after upgrade success"
+}
+
+// Setup creates a service with a load balancer and makes sure it's reachable.
+func (t *UpgradeTest) Setup(f *framework.Framework) {
+	g.By("Setting up post-upgrade alert test")
+	oc := exutil.NewCLIWithoutNamespace("prometheus")
+	var ok bool
+	url, bearerToken, ok := locatePrometheus(oc)
+	if !ok {
+		framework.Failf("Prometheus could not be located on this cluster, failing upgrade alert test")
+	}
+	t.url = url
+	t.bearerToken = bearerToken
+	t.oc = oc
+	framework.Logf("Post-upgrade alert test setup complete")
+}
+
+// Test runs a connectivity check to the service.
+func (t *UpgradeTest) Test(f *framework.Framework, done <-chan struct{}, upgrade upgrades.UpgradeType) {
+
+	// Recover current test if it fails so test suite can complete
+	defer g.GinkgoRecover()
+
+	// Block until upgrade is done
+	g.By("Waiting for upgrade to finish before checking for critical alerts")
+	<-done
+
+	_, cancel := context.WithCancel(context.Background())
+	start := time.Now()
+
+	g.By("Waiting before checking for critical alerts")
+	time.Sleep(alertCheckSleep)
```

Comment:
out of curiosity, why do you have to sleep here?

Reply:
This change is to satisfy [OTA-141](https://issues.redhat.com/browse/OTA-141) whereby we want to check if critical alerts are present on cluster post-upgrade after some period. So we want to give the cluster a chance to settle/run steady state after the upgrade and then check for any critical alerts.

Reply:
Would be good to capture this motivation^ in the commit message.

Reply:
This commit was not to specific to this comment. The reason I have so many commits is that I was doing all my testing through CI (becuase I couldn't get the upgrade e2e test to run myself) so Iwas hacking around, testing, developing, etc. The sleep was always going to be there as part of the overall design of the product

Reply:
> The sleep was always going to be there as part of the overall design of the product.

OTA-141 is not public (at the moment).  And even for folks who do have access, it's a separate click out of `git blame ...` beyond the commit that landed the code they're wondering about.  So it's nice to outline the design in the commit that lands the implementation so it's conveniently close for anyone poking about in the public Git repo.